programs/perpetuals/src/instructions/resolve_locked_stake.rs

```diff
@@ -0,0 +1,196 @@
+        error::PerpetualsError,
+        math, program,
+        state::{cortex::Cortex, perpetuals::Perpetuals, staking::Staking},
+    },
+    anchor_lang::prelude::*,
+    anchor_spl::token::{Mint, Token},
+};
+
+#[derive(Accounts)]
+pub struct ResolveLockedStake<'info> {
+    // Caller is restrained to be the clockwork thread only
+    #[account(mut, owner = clockwork_sdk::ID)]
+    pub caller: Signer<'info>,
+
+    /// CHECK: verified through the `stake` account seed derivation
+    #[account(mut)]
+    pub owner: AccountInfo<'info>,
+
+    /// CHECK: empty PDA, authority for token accounts
+    #[account(
+        seeds = [b"transfer_authority"],
+        bump = perpetuals.transfer_authority_bump
+    )]
+    pub transfer_authority: AccountInfo<'info>,
+
+    #[account(
+        mut,
+        seeds = [b"staking",
+                 owner.key().as_ref()],
+        bump = staking.bump
+    )]
+    pub staking: Box<Account<'info, Staking>>,
+
+    #[account(
+        mut,
+        seeds = [b"cortex"],
+        bump = cortex.bump,
+    )]
+    pub cortex: Box<Account<'info, Cortex>>,
+
+    #[account(
+        seeds = [b"perpetuals"],
+        bump = perpetuals.perpetuals_bump
+    )]
+    pub perpetuals: Box<Account<'info, Perpetuals>>,
+
+    #[account(
+        mut,
+        seeds = [b"lm_token_mint"],
+        bump = cortex.lm_token_bump
+    )]
+    pub lm_token_mint: Box<Account<'info, Mint>>,
+
+    #[account(
+        mut,
+        seeds = [b"governance_token_mint"],
+        bump = cortex.governance_token_bump
+    )]
+    pub governance_token_mint: Box<Account<'info, Mint>>,
+
+    /// CHECK: checked by spl governance v3 program
+    /// A realm represent one project (ADRENA, MANGO etc.) within the governance program
+    pub governance_realm: UncheckedAccount<'info>,
+
+    /// CHECK: checked by spl governance v3 program
+    pub governance_realm_config: UncheckedAccount<'info>,
+
+    /// CHECK: checked by spl governance v3 program
+    /// Token account owned by governance program holding user's locked tokens
+    #[account(mut)]
+    pub governance_governing_token_holding: UncheckedAccount<'info>,
+
+    /// CHECK: checked by spl governance v3 program
+    /// Account owned by governance storing user informations
+    #[account(mut)]
+    pub governance_governing_token_owner_record: UncheckedAccount<'info>,
+    governance_program: Program<'info, SplGovernanceV3Adapter>,
+    perpetuals_program: Program<'info, program::Perpetuals>,
+    system_program: Program<'info, System>,
+    token_program: Program<'info, Token>,
+}
+
+#[derive(AnchorSerialize, AnchorDeserialize, Copy, Clone)]
+pub struct ResolveLockedStakeParams {
+    pub thread_id: u64,
+}
+
+// Resolving a stake means cancelling the governing power related to the stake and stopping to accrue rewards
+// A stake can be resolved when its locking period have ended
+// After a stake is resolved, it can be removed by the user to retrieve its tokens
```

Comment:
👍

Reply:
maybe finalize_locked_stake (resolve word used for rounds already. But okish to reuse), and claim_locked_stake instead of remove? The remove is a side effect of the claim, but it being called remove feels like you'r altering the locked stake before completion? (maybe just me, also just naming)